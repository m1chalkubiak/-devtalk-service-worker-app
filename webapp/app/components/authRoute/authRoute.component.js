import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Redirect, Route } from 'react-router-dom';


export class AuthRoute extends Component {
  static propTypes = {
    isAuthenticated: PropTypes.bool.isRequired,
    anonymous: PropTypes.bool.isRequired,
    defaultRedirect: PropTypes.string.isRequired,
    location: PropTypes.object.isRequired,
  };

  static defaultProps = {
    anonymous: false,
    defaultRedirect: '/login',
  };

  renderRoute = () => <Route {...this.props} />;

  render() {
    const { isAuthenticated, anonymous, defaultRedirect } = this.props;

    if (isAuthenticated) {
      /**
       * This route is only viewable by unauthenticated users
       */
      if (anonymous) {
        return <Redirect to="/" />;
      }

      /**
       * User has met all criteria and the route can be rendered
       */
      return this.renderRoute();
    }

    /**
     * This route is viewable by anonymous users
     */
    if (anonymous) {
      return this.renderRoute();
    }

    /**
     * Redirect to register by default
     */
    return <Redirect to={defaultRedirect} />;
  }
}
