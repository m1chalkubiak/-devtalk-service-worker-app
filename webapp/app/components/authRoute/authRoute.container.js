import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { createStructuredSelector } from 'reselect';
import { compose } from 'ramda';
import { hot } from 'react-hot-loader';

import { selectIsConfigured } from '../../modules/userAuth';
import { AuthRoute } from './authRoute.component';


const mapStateToProps = createStructuredSelector({
  isAuthenticated: selectIsConfigured,
});

export const mapDispatchToProps = (dispatch) => bindActionCreators({}, dispatch);

export default compose(
  hot(module),
  connect(mapStateToProps, mapDispatchToProps)
)(AuthRoute);
