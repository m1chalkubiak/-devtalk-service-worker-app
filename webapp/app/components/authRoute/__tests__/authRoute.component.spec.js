import React from 'react';
import { shallow } from 'enzyme';
import { Map } from 'immutable';

import { AuthRoute } from '../authRoute.component';

describe('<AuthRoute />', () => {
  const defaultProps = {
    isAuthenticated: false,
    userProfile: Map(),
    location: {},
    hasInstalledApp: false,
  };

  const component = (props) => (
    <AuthRoute to="/some-route" {...defaultProps} {...props} />
  );

  it('should render default component tree', () => {
    const wrapper = shallow(component());
    global.expect(wrapper).toMatchSnapshot();
  });

  it('should render a <Redirect /> to home page when user is authenticated but route requires anonymous', () => {
    const wrapper = shallow(component({ anonymous: true, isAuthenticated: true }));
    global.expect(wrapper).toMatchSnapshot();
  });

  it('should render a <Redirect /> to login page if user is not authenticated', () => {
    const wrapper = shallow(component());
    global.expect(wrapper).toMatchSnapshot();
  });

  it('should render null when user is authenticated by user roles were not fetched yet', () => {
    const wrapper = shallow(component({ isAuthenticated: true }));
    global.expect(wrapper).toMatchSnapshot();
  });

  it('should render a <Redirect /> to login page if user is archived', () => {
    const wrapper = shallow(component({
      userProfile: Map({
        name: 'John Doe',
        archived: true,
      }),
      isAuthenticated: true,
    }));
    global.expect(wrapper).toMatchSnapshot();
  });

  it('should render a real <Route /> when user is authenticated and is not archived', () => {
    const wrapper = shallow(component({
      userProfile: Map({
        name: 'John Doe',
      }),
      isAuthenticated: true,
    }));
    global.expect(wrapper).toMatchSnapshot();
  });

  it('should render a real <Route /> when user is not authenticated and route required anonymous user', () => {
    const wrapper = shallow(component({
      isAuthenticated: false,
      anonymous: true,
    }));
    global.expect(wrapper).toMatchSnapshot();
  });

  it('should render a <Redirect /> to defaultRedirect page', () => {
    const wrapper = shallow(component({
      isAuthenticated: false,
      anonymous: false,
      defaultRedirect: '/default-page',
    }));
    global.expect(wrapper).toMatchSnapshot();
  });
});
