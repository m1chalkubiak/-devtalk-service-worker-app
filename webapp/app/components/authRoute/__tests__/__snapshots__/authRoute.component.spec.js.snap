// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AuthRoute /> should render a <Redirect /> to defaultRedirect page 1`] = `
<Redirect
  push={false}
  to="/default-page"
/>
`;

exports[`<AuthRoute /> should render a <Redirect /> to home page when user is authenticated but route requires anonymous 1`] = `
<Redirect
  push={false}
  to="/"
/>
`;

exports[`<AuthRoute /> should render a <Redirect /> to login page if user is archived 1`] = `
<Route
  anonymous={false}
  defaultRedirect="/login"
  hasInstalledApp={false}
  isAuthenticated={true}
  location={Object {}}
  to="/some-route"
  userProfile={
    Immutable.Map {
      "name": "John Doe",
      "archived": true,
    }
  }
/>
`;

exports[`<AuthRoute /> should render a <Redirect /> to login page if user is not authenticated 1`] = `
<Redirect
  push={false}
  to="/login"
/>
`;

exports[`<AuthRoute /> should render a real <Route /> when user is authenticated and is not archived 1`] = `
<Route
  anonymous={false}
  defaultRedirect="/login"
  hasInstalledApp={false}
  isAuthenticated={true}
  location={Object {}}
  to="/some-route"
  userProfile={
    Immutable.Map {
      "name": "John Doe",
    }
  }
/>
`;

exports[`<AuthRoute /> should render a real <Route /> when user is not authenticated and route required anonymous user 1`] = `
<Route
  anonymous={true}
  defaultRedirect="/login"
  hasInstalledApp={false}
  isAuthenticated={false}
  location={Object {}}
  to="/some-route"
  userProfile={Immutable.Map {}}
/>
`;

exports[`<AuthRoute /> should render default component tree 1`] = `
<Redirect
  push={false}
  to="/login"
/>
`;

exports[`<AuthRoute /> should render null when user is authenticated by user roles were not fetched yet 1`] = `
<Route
  anonymous={false}
  defaultRedirect="/login"
  hasInstalledApp={false}
  isAuthenticated={true}
  location={Object {}}
  to="/some-route"
  userProfile={Immutable.Map {}}
/>
`;
